File: leetcode/783-二叉搜索树节点最小距离/783-二叉搜索树节点最小距离.go
package main

import "math"

/**
 * Definition for a binary tree node.
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 783-二叉搜索树节点最小距离
func minDiffInBST(root *TreeNode) int {
	if root == nil {
		return 0
	}
	minVal := math.MaxInt64
	pre := -1
	var traverse func(*TreeNode)
	traverse = func(root *TreeNode) {
		if root == nil {
			return
		}
		traverse(root.Left)
		if pre != -1 && root.Val-pre < minVal {
			minVal = root.Val - pre
		}
		pre = root.Val
		traverse(root.Right)
	}
	traverse(root)
	return minVal
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

File: leetcode/783-二叉搜索树节点最小距离/783-二叉搜索树节点最小距离.py
# Definition for a binary tree node.
from re import L


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
# 783-二叉搜索树节点最小距离
class Solution:
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        minVal = 100001
        pre = -1
        def traverse(root):
            if (root is None):
                return
            traverse(root.left)
            nonlocal pre,minVal
            if pre != -1 and root.val - pre < minVal:
                minVal = root.val - pre
            pre = root.val
            traverse(root.right)
        traverse(root)
        return minVal